File: Python/bessel.py
import numpy as np
import matplotlib.pyplot as plt


def besselj_list(n_max, x):
    n = n_max - 2
    j = [1.0, 0.0]
    scale = 0.0
    x = float(x)

    if x == 0.0:
        return [1] + (n_max - 1) * [0]
    

    else:
        while n >= 0:
            j = [2.0 * float(n + 1) * j[0] / x - j[1]] + j
            if n == 0:
                scale = scale + j[0]
            elif n % 2 == 0:
                scale = scale + 2.0 * j[0]

            n = n - 1

        j = [val / scale for val in j]

        return j

# print(besselj_list(60, 22026.465749406787))

# def besselj(n, x):
#     if x == 0.0:
#         if n == 0:
#             val = 1.0
#         else:
#             val = 0.0
#     else:
#         if x < n ** 2:
#             val = besselj_list(3 * n + 6 * np.ceil(x), x)[n]
#         else:
#             val = np.sqrt(2 / (np.pi * x)) * np.cos(x - (2.0 * np.pi + 1.0) / 4.0)

#     return val

def besselj(n, x):
    if x == 0.0:
        if n == 0:
            val = 1.0
        else:
            val = 0.0
    elif 0.0 < x < np.sqrt(n + 1.0):
        val = ((x / 2.0) ** n) * (1.0 / np.math.factorial(n))
    else:
        val = np.sqrt(2.0 / (np.pi * x)) * np.cos(x - (2.0 * n * np.pi + 1.0) / 4.0)

    return val

# print([besselj(0, float(i) / 10) for i in range(500)])

# t = np.linspace(0.0, 50.0, 400)
# bess = []
# for val in t:
#     bess.append(besselj(2, val))

# print(bess)
# plt.figure()
# plt.plot(t, bess, 'k-')
# plt.grid(True)
# plt.show()

File: js/bessel.js
function besselj_list(n_max, x) {

    var n = n_max - 2;
    var j = [1, 0];
    var scale = 0;
    var unitCoeff = [];

    if (x == 0) {

        var list = [1];

        for (var i = 0; i <= n; i++) {

            list.push(0);

        }

        return list;

    } else if (x < n_max / 2) {

        while (n >= 0) {

            var j1 = [];
            var val = 2 * (n + 1) * j[0] / x - j[1];
            j.forEach((x) => { j1.push(x / val) });
            j = [1].concat(j1);
            n--;

        }

        for (var i = 0; i < n_max; i++) {

            !(i % 2) ? scale += 2 * j[i] : scale += 0;

        }

        scale -= j[0];

        for (var i = 0; i < n_max; i++) {

            unitCoeff.push(j[i] / scale);

        }

        return unitCoeff;

    } else {

        for (var i = 0; i < n_max; i++) {

            unitCoeff.push(besselj_asypm(x, i, 5));

        }

        return unitCoeff;

    }

}

function besselj_asypm(x, n, order) {

    const w = x - n * Math.PI / 2 - Math.PI / 4;
    var cosTerm = 0, sinTerm = 0;

    for (var k = 0; k <= order; k++) {

        cosTerm += ((-1) ** k) * a(2 * k, n) / (x ** (2 * k));
        sinTerm += ((-1) ** k) * a(2 * k + 1, n) / (x ** (2 * k + 1));

    }

    cosTerm *= Math.cos(w);
    sinTerm *= Math.sin(w);

    var j = cosTerm - sinTerm;
    j *= Math.sqrt(2 / (Math.PI * x));

    return j;

}

function besselj(x, n, scale = 1) {

    const accuracy = 50;
    const asympOrder = 6;

    if (typeof (x) === "number") {

        if (x < n) {

            return besselj_list(accuracy, scale * x)[n];

        } else {

            return besselj_asypm(x, n, asympOrder);

        }

    } else {

        var vals = [];

        x.forEach((x_val) => {

            if (x_val < accuracy) {

                vals.push(besselj_list(accuracy, scale * x_val)[n]);

            } else {

                vals.push(besselj_asypm(x_val, n, asympOrder));

            }

        });

        return vals;

    }

}

function a(k, v) {

    var a = 1;
    var n = 1;

    while (n <= 2 * k - 1) {

        a *= (2 * v - n) * (2 * n + n) / (8 * n);
        n += 2;

    }

    return a

}

export { besselj_list, besselj };